demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb

```diff
@@ -0,0 +1,631 @@
+    "        messages_history += dialog\n",
+    "        \n",
+    "        #Queue for streamed character rendering\n",
+    "        q = Queue()\n",
+    "\n",
+    "        #Update new llama hyperparameters\n",
+    "        llm.inference_server_url = model_selector\n",
+    "        llm.temperature = temperature_selector\n",
+    "        llm.top_p = top_p_selector\n",
+    "        llm.max_new_tokens = max_new_tokens_selector\n",
+    "\n",
+    "        #Async task for streamed chain results wired to callbacks we previously defined, so we don't block the UI\n",
+    "        async def task(prompt):\n",
+    "            ret = await qa_chain.run(prompt, callbacks=[MyStream(q)])\n",
+    "            return ret\n",
+    "\n",
+    "        with start_blocking_portal() as portal:\n",
+    "            portal.start_task_soon(task, user_prompt_message)\n",
+    "            while True:\n",
+    "                next_token = q.get(True)\n",
+    "                if next_token is job_done:\n",
+    "                    messages_history += [{\"role\": \"assistant\", \"content\": bot_message}]\n",
+    "                    return history, messages_history\n",
+    "                bot_message += next_token\n",
+    "                history[-1][1] += next_token\n",
+    "                yield history, messages_history\n",
+    "\n",
+    "    #init the chat history with default system message    \n",
+    "    def init_history(messages_history):\n",
+    "        messages_history = []\n",
+    "        messages_history += [system_message]\n",
+    "        return messages_history\n",
+    "\n",
+    "    #clean up the user input text field\n",
+    "    def input_cleanup():\n",
+    "        return \"\"\n",
+    "\n",
+    "    #when user click Enter and user message are submitted\n",
+    "    user_prompt_message.submit(user, [user_prompt_message, chatbot], [chatbot], queue=False).then(\n",
+    "        bot, [model_selector, temperature_selector, top_p_selector, max_new_tokens_selector, user_prompt_message, chatbot, state], [chatbot, state]\n",
+    "    ).then(input_cleanup, [], [user_prompt_message], queue=False)\n",
+    "\n",
+    "    #when user click the submit button\n",
+    "    submitBtn.click(user, [user_prompt_message, chatbot], [chatbot], queue=False).then(\n",
+    "        bot, [model_selector, temperature_selector, top_p_selector, max_new_tokens_selector, user_prompt_message, chatbot, state], [chatbot, state]\n",
+    "    ).then(input_cleanup, [], [user_prompt_message], queue=False)\n",
+    "    \n",
+    "    #when user click the clear button\n",
+    "    clear.click(lambda: None, None, chatbot, queue=False).success(init_history, [state], [state])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Lastly, we can launch this demo on our localhost. Make sure you have access to the ports.  \n",
+    "In the notebook or a browser with URL http://0.0.0.0:7860 you should see the UI.  \n",
```

Comment:
when can one open the URL http://0.0.0.0:7860? after running the following code?
demo.queue().launch(server_name="0.0.0.0")

also what's the port 7860? the default one?

A screenshot of the UI will be nice.

Reply:
Rearragne the narrative to make it clear. 
I don't think a screenshot of UI is necessary though. You either launch it successfully or you can't reach anything in that port.

Reply:
Just went through your whole notebook and was able to run the demo and see the Gradio UI both from the notebook and the browser with the URL. Maybe a better way than a screenshot, to make the demo more interesting  on a quick read (without having to run it), is to save the cell output like this:

![image](https://github.com/facebookresearch/llama-recipes/assets/535090/22c5937a-30d6-4897-9853-8f866b6c053f)

Three other comments:

1. About "Things to try in the chatbot demo:

Asking specific questions related to the Llama 2 Getting Started Guide
Streaming
Adjust parameters such as max new token generated"

Since "streaming" is not an element in the UI, is modifying it in the `llm = HuggingFaceTextGenInference(...)` call the only way to try it out? If so, maybe we don't list it here.

2. We probably should replace the code cell with

model = meta-llama/Llama-2-7b-chat-hf  
volume = $PWD/data  
token = #Your own HF tokens  
docker run --gpus all --shm-size 1g -e HUGGING_FACE_HUB_TOKEN=$token -p 8080:80 -v $volume:/data ghcr.io/huggingface/text-generation-inference:1.1.0 --model-id $model

with a Markdown cell and use ``` ... ``` in the cell, otherwise the user may think it's the same as other code cells in the notebook, run it and get an error (even though you mentioned "on a separate terminal".

3. The user may likely run the notebook on a separate machine than the machine running TGI, then using localhost will fail. So you may want to add a note asking them to replace "localhost" in the 3 cases with the IP visible to the machine running the notebook.
 
!curl localhost:8080...
LLAMA2_7B_HOSTPORT = "http://localhost:8080/"
LLAMA2_13B_HOSTPORT = "http://localhost:8080/" #Add your own host ports for model switching. You can host another TGI model on same instance on a different port.

Reply:
Thanks @jeffxtang, accept all 4 comments/suggestions and updated.